tensorflow_probability/python/experimental/mcmc/rhat_reducer_test.py

```diff
@@ -0,0 +1,206 @@
+    true_rhat = tfp.mcmc.potential_scale_reduction(
+        chains_states=chain_state,
+        independent_chain_ndims=1,
+    )
+    self.assertEqual(tf.float64, rhat.dtype)
+    rhat, true_rhat = self.evaluate([rhat, true_rhat])
+    self.assertAllClose(true_rhat, rhat, rtol=1e-6)
+
+  def test_chunking(self):
+    rhat_reducer = tfp.experimental.mcmc.RhatReducer(
+        num_parallel_chains=5,
+    )
+    state = rhat_reducer.initialize(tf.zeros(()))
+    chain_state = np.arange(60, dtype=np.float32).reshape((4, 5, 3))
+    for sample in chain_state:
+      state = rhat_reducer.one_step(sample, state, axis=0)
+    rhat = rhat_reducer.finalize(state)
+    true_rhat = tfp.mcmc.potential_scale_reduction(
+        chains_states=np.swapaxes(chain_state, 1, 2).reshape((12, 5)),
+        independent_chain_ndims=1,
+    )
+    rhat, true_rhat = self.evaluate([rhat, true_rhat])
+    self.assertAllClose(true_rhat, rhat, rtol=1e-6)
+
+  def test_in_with_reductions(self):
+    rhat_reducer = tfp.experimental.mcmc.RhatReducer(
+        num_parallel_chains=5,
+    )
+    fake_kernel = TestTransitionKernel(shape=(5,))
+    reduced_kernel = tfp.experimental.mcmc.WithReductions(
+        inner_kernel=fake_kernel,
+        reducer=rhat_reducer,
+    )
+    chain_state = tf.zeros(5,)
+    pkr = reduced_kernel.bootstrap_results(chain_state)
+    for _ in range(2):
+      chain_state, pkr = reduced_kernel.one_step(
+          chain_state, pkr)
+    rhat = self.evaluate(
+        rhat_reducer.finalize(pkr.streaming_calculations))
+    self.assertEqual(0.5, rhat)
+
+  def test_iid_normal_passes(self):
+    n_samples = 500
+    # two scalar chains taken from iid Normal(0, 1)
+    rng = test_util.test_np_rng()
+    iid_normal_samples = rng.randn(n_samples, 2)
+    rhat_reducer = tfp.experimental.mcmc.RhatReducer(
+        num_parallel_chains=2,
+    )
+    state = rhat_reducer.initialize(iid_normal_samples[0])
+    for sample in iid_normal_samples:
+      state = rhat_reducer.one_step(sample, state)
+    rhat = self.evaluate(rhat_reducer.finalize(state))
+    self.assertAllEqual((), rhat.shape)
+    self.assertAllClose(1., rhat, rtol=0.02)
+
+  def test_offset_normal_fails(self):
+    n_samples = 500
+    # three 4-variate chains taken from Normal(0, 1) that have been
+    # shifted. Since every chain is shifted, they are not the same, and the
+    # test should fail.
+    offset = np.array([1., -1., 2.]).reshape(3, 1)
+    rng = test_util.test_np_rng()
+    offset_samples = rng.randn(n_samples, 3, 4) + offset
+    rhat_reducer = tfp.experimental.mcmc.RhatReducer(
+        num_parallel_chains=3,
+    )
+    state = rhat_reducer.initialize(offset_samples[0])
+    for sample in offset_samples:
+      state = rhat_reducer.one_step(sample, state)
+    rhat = self.evaluate(rhat_reducer.finalize(state))
+    self.assertAllEqual((4,), rhat.shape)
+    self.assertAllEqual(np.ones_like(rhat).astype(bool), rhat > 1.2)
+
+  def test_with_hmc(self):
```

Comment:
Can we also add a test with more than one latent Tensor, following some of the tests of `tfp.mcmc.potential_scale_reduction_factor`?

Reply:
independent R-hat calculations on multiple states